Add new rule `no-down-event-binding`

alexlafroscia opened the pull request:
This PR adds a new lint rule that catches usage of "down" events in the `{{on}}` or `{{action}}` modifier and guides developers to use the "up" versions of them instead. This is not _always_ the right thing to do, but is _almost_ always the right thing to do, so I think it's worth having a lint rule for!

This came up thanks to a Tweet from @MelSumner today, where she linked to the WCAG guideline around not using "down" events. This was a major TIL for me, and since I'd bet most developers just guess which even they should be using, it felt like a lint rule could really help push people into the proverbial "pit of success"!

https://twitter.com/melaniersumner/status/1330905297326649344

alexlafroscia commented:
By the way, I just wanted to say that the experience of adding new rules here is _so good_. I love that the contribution docs walk through exactly how to generate the rule, run tests, update the README, etc. It really made creating this a piece of cake!

lifeart commented on lib/rules/no-down-event-binding.js:
```diff
@@ -0,0 +1,74 @@
+      ElementModifierStatement(node) {
+        if (isOnModifier(node)) {
+          const eventNameNode = node.params[0];
+
```
Could we have fixer for this rule? 
Looks like it’s a simple one

alexlafroscia replied:
It would definitely be easy to write! My only concern is the idea of auto-fixing something that is really different behavior. I don't know what I would want to fix this on save and not notice that things had changed, in case you really *do* need to bind to a "down" event..

bmish replied:
Autofixers should not change behavior. Changing the event used could be a significant change in behavior.

bmish commented on test/unit/rules/no-down-event-binding-test.js:
```diff
@@ -0,0 +1,46 @@
+    "<div {{action this.doSomething on='keyup'}}></div>",
+  ],
+
+  bad: [
```
What about the `onkeydown` DOM event?

https://www.w3schools.com/jsref/event_onkeydown.asp

alexlafroscia replied:
Yeah, that would be caught by this -- the tests aren't necessarily exhaustive. Would you prefer I add test cases for that event specifically? This rule disallows any `__down` event.

bmish replied:
I don't see any test case with pure HTML. Can you add this?

```html
<input type="text" onkeydown="myFunction()">
```

alexlafroscia replied:
Oh! Great point. I'll add a test for that!

bmish commented on test/unit/rules/no-down-event-binding-test.js:
```diff
@@ -0,0 +1,46 @@
+    "<div {{action this.doSomething on='keyup'}}></div>",
+  ],
+
+  bad: [
```
What about components?

```hbs
<MyComponent @keyDown={{@close}}>
```

```hbs
{{my-component keyDown=foo}}
```

alexlafroscia replied:
Hmm, that's an interesting idea. That could handle "classic component" arguments that automatically become event listeners, but that's increasingly irrelevant for Glimmer components that don't use that behavior. 

If we think about the mental model going forward and assume these arguments do not automatically become event handlers, would we want to lint random component arguments? It would probably be safe to assume that an argument called `keyDown` maps to such an event listener, but would that be assuming too much from the linter's perspective?

bmish replied:
I'd be fine excluding components as the default behavior. If someone wanted to follow-up later, they could add a rule option `catchComponents` (default false) to be more aggressive. 

For now, could you add these as valid test cases then?

alexlafroscia replied:
Sure, I'll add those tests, too!

alexlafroscia commented:
Heads-up: I rebased one more time. I realized that when I updated the behavior and tests based on the feedback from @bmish, I forgot to update the documentation file! That's been fixed now.
